scapy/layers/ppp.py

```diff
@@ -552,18 +552,151 @@ class PPP_LCP_Discard_Request(PPP_LCP):
                    IntField("magic_number", None),
                    StrLenField("data", "", length_from=lambda p:p.len-8)]
 
+### Password authentication protocol (RFC 1334)
+
+_PPP_paptypes = {1: "Authenticate-Request",
+                 2: "Authenticate-Ack",
+                 3: "Authenticate-Nak"}
+
+
+class PPP_PAP(Packet):
+    name = "PPP Password Authentication Protocol"
+    fields_desc = [ByteEnumField("code", 1, _PPP_paptypes),
+                   XByteField("id",0),
+                   ShortField("len", None),
+                   StrLenField("data","",length_from=lambda p:p.len-4)]
+
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *args, **kargs):
+        if _pkt:
+            o = ord(_pkt[0])
+            if o == 1:
+                return PPP_PAP_Request
+            elif o in [2,3]:
+                return PPP_PAP_Response
+        return cls
+
+    def extract_padding(self, pay):
+        return "",pay
+
+    def post_build(self, p, pay):
+        p += pay
+        if self.len is None:
+            p = p[:2]+struct.pack("!H", len(p))+p[4:]
+        return p
+
+
+class PPP_PAP_Request(PPP_PAP):
+    fields_desc = [ByteEnumField("code", 1, _PPP_paptypes),
+                   XByteField("id", 0),
+                   ShortField("len", None),
+                   FieldLenField("username_len", None, fmt="B", length_of="username"),
+                   StrLenField("username", None, length_from=lambda p: p.username_len),
+                   FieldLenField("passwd_len", None, fmt="B", length_of="password"),
+                   StrLenField("password", None, length_from=lambda p: p.passwd_len)]
+
+    def mysummary(self):
+        return "PAP-Request username='{0}' password='{1}'".format(self.username, self.password)
+
+
+class PPP_PAP_Response(PPP_PAP):
+    fields_desc = [ByteEnumField("code", 2, _PPP_paptypes),
+                   XByteField("id", 0),
+                   ShortField("len", None),
+                   FieldLenField("msg_len", None, fmt="B", length_of="message"),
+                   StrLenField("message", "", length_from=lambda p: p.msg_len)]
+
+    def answers(self, other):
+        return isinstance(other, PPP_PAP_Request) and other.id == self.id
+
+    def mysummary(self):
+        res = "PAP-Ack" if self.code == 2 else "PAP-Nak"
+        if self.msg_len > 0:
+            res += " msg='{0}'".format(self.message)
```

Comment:
Same remark here (and in the next `.mysummary()` implementations) about using `sprintf()`.